File: sdk/python/aragora_sdk/namespaces/system.py
"""
System Namespace API.

Provides system administration and monitoring:
- Database maintenance tasks
- History tracking (cycles, events, debates)
- Circuit breaker monitoring
- Authentication statistics
- Debug endpoints
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

MaintenanceTask = Literal["status", "vacuum", "analyze", "checkpoint", "full"]
CircuitBreakerStatus = Literal["closed", "open", "half_open"]


class SystemAPI:
    """
    Synchronous System API.

    Provides methods for system administration:
    - Database maintenance tasks
    - History tracking (cycles, events, debates)
    - Circuit breaker monitoring
    - Authentication statistics
    - Debug endpoints

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> # Run database maintenance
        >>> result = client.system.run_maintenance("vacuum")
        >>> # Get circuit breaker status
        >>> breakers = client.system.get_circuit_breakers()
        >>> # Get debate history
        >>> history = client.system.get_debate_history(limit=50)
    """

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    def debug_test(self) -> dict[str, Any]:
        """
        Run a debug test endpoint.

        Useful for verifying API connectivity.

        Returns:
            Debug test response with status and method.
        """
        return self._client.request("GET", "/api/debug/test")

    def get_cycles(
        self,
        loop_id: str | None = None,
        limit: int = 50,
    ) -> dict[str, Any]:
        """
        Get cycle history.

        Args:
            loop_id: Optional filter by loop ID.
            limit: Maximum number of results.

        Returns:
            List of cycle entries.
        """
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return self._client.request("GET", "/api/history/cycles", params=params)

    def get_events(
        self,
        loop_id: str | None = None,
        limit: int = 100,
    ) -> dict[str, Any]:
        """
        Get event history.

        Args:
            loop_id: Optional filter by loop ID.
            limit: Maximum number of results.

        Returns:
            List of event entries.
        """
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return self._client.request("GET", "/api/history/events", params=params)

    def get_debate_history(
        self,
        loop_id: str | None = None,
        limit: int = 50,
    ) -> dict[str, Any]:
        """
        Get debate history.

        Args:
            loop_id: Optional filter by loop ID.
            limit: Maximum number of results.

        Returns:
            List of debate history entries.
        """
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return self._client.request("GET", "/api/history/debates", params=params)

    def get_history_summary(self, loop_id: str | None = None) -> dict[str, Any]:
        """
        Get history summary statistics.

        Args:
            loop_id: Optional filter by loop ID.

        Returns:
            Summary with total debates, agents, matches.
        """
        params: dict[str, Any] = {}
        if loop_id:
            params["loop_id"] = loop_id
        return self._client.request("GET", "/api/history/summary", params=params)

    def run_maintenance(
        self,
        task: MaintenanceTask = "status",
    ) -> dict[str, Any]:
        """
        Run database maintenance task.

        Args:
            task: Type of maintenance to run.

        Returns:
            Maintenance result with success status.
        """
        return self._client.request("GET", "/api/system/maintenance", params={"task": task})

    def get_auth_stats(self) -> dict[str, Any]:
        """
        Get authentication statistics.

        Returns:
            Auth stats with user counts, sessions, tokens.
        """
        return self._client.request("GET", "/api/auth/stats")

    def revoke_token(
        self,
        token_id: str | None = None,
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """
        Revoke a token.

        Args:
            token_id: Token identifier to revoke.
            user_id: User ID to revoke tokens for.

        Returns:
            Revocation confirmation.
        """
        data: dict[str, Any] = {}
        if token_id:
            data["token_id"] = token_id
        if user_id:
            data["user_id"] = user_id
        return self._client.request("POST", "/api/auth/revoke", json=data)

    def get_circuit_breakers(self) -> dict[str, Any]:
        """
        Get circuit breaker metrics.

        Returns:
            Dictionary of circuit breaker names to metrics.
        """
        return self._client.request("GET", "/api/circuit-breakers")

    def get_prometheus_metrics(self) -> dict[str, Any]:
        """
        Get Prometheus metrics.

        Returns:
            Prometheus format metrics.
        """
        return self._client.request("GET", "/metrics")

    def get_system_intelligence_overview(self) -> dict[str, Any]:
        """Get high-level system intelligence dashboard stats."""
        return self._client.request("GET", "/api/v1/system-intelligence/overview")

    def get_system_intelligence_agent_performance(self) -> dict[str, Any]:
        """Get agent ELO, calibration, and win-rate dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/agent-performance")

    def get_system_intelligence_institutional_memory(self) -> dict[str, Any]:
        """Get institutional-memory dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/institutional-memory")

    def get_system_intelligence_improvement_queue(self) -> dict[str, Any]:
        """Get improvement-queue dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/improvement-queue")

    def get_system_intelligence_anomalies(self) -> dict[str, Any]:
        """Get recent anomaly alerts for the system-intelligence dashboard."""
        return self._client.request("GET", "/api/v1/system-intelligence/anomalies")

    def get_system_intelligence_events(self, limit: int | None = None) -> dict[str, Any]:
        """Get recent system events for the system-intelligence dashboard."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return self._client.request("GET", "/api/v1/system-intelligence/events", params=params)

    def get_system_intelligence_km_sync(self) -> dict[str, Any]:
        """Get Knowledge Mound sync dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/km-sync")

    def get_system_intelligence_nomic_status(self) -> dict[str, Any]:
        """Get nomic loop status dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/nomic-status")

    def get_system_intelligence_debate_queue(self) -> dict[str, Any]:
        """Get debate queue activity dashboard data."""
        return self._client.request("GET", "/api/v1/system-intelligence/debate-queue")


class AsyncSystemAPI:
    """Asynchronous System API."""

    def __init__(self, client: AragoraAsyncClient) -> None:
        self._client = client

    async def debug_test(self) -> dict[str, Any]:
        """Run a debug test endpoint."""
        return await self._client.request("GET", "/api/debug/test")

    async def get_cycles(
        self,
        loop_id: str | None = None,
        limit: int = 50,
    ) -> dict[str, Any]:
        """Get cycle history."""
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return await self._client.request("GET", "/api/history/cycles", params=params)

    async def get_events(
        self,
        loop_id: str | None = None,
        limit: int = 100,
    ) -> dict[str, Any]:
        """Get event history."""
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return await self._client.request("GET", "/api/history/events", params=params)

    async def get_debate_history(
        self,
        loop_id: str | None = None,
        limit: int = 50,
    ) -> dict[str, Any]:
        """Get debate history."""
        params: dict[str, Any] = {"limit": limit}
        if loop_id:
            params["loop_id"] = loop_id
        return await self._client.request("GET", "/api/history/debates", params=params)

    async def get_history_summary(self, loop_id: str | None = None) -> dict[str, Any]:
        """Get history summary statistics."""
        params: dict[str, Any] = {}
        if loop_id:
            params["loop_id"] = loop_id
        return await self._client.request("GET", "/api/history/summary", params=params)

    async def run_maintenance(
        self,
        task: MaintenanceTask = "status",
    ) -> dict[str, Any]:
        """Run database maintenance task."""
        return await self._client.request("GET", "/api/system/maintenance", params={"task": task})

    async def get_auth_stats(self) -> dict[str, Any]:
        """Get authentication statistics."""
        return await self._client.request("GET", "/api/auth/stats")

    async def revoke_token(
        self,
        token_id: str | None = None,
        user_id: str | None = None,
    ) -> dict[str, Any]:
        """Revoke a token."""
        data: dict[str, Any] = {}
        if token_id:
            data["token_id"] = token_id
        if user_id:
            data["user_id"] = user_id
        return await self._client.request("POST", "/api/auth/revoke", json=data)

    async def get_circuit_breakers(self) -> dict[str, Any]:
        """Get circuit breaker metrics."""
        return await self._client.request("GET", "/api/circuit-breakers")

    async def get_prometheus_metrics(self) -> dict[str, Any]:
        """Get Prometheus metrics."""
        return await self._client.request("GET", "/metrics")

    async def get_system_intelligence_overview(self) -> dict[str, Any]:
        """Get high-level system intelligence dashboard stats."""
        return await self._client.request("GET", "/api/v1/system-intelligence/overview")

    async def get_system_intelligence_agent_performance(self) -> dict[str, Any]:
        """Get agent ELO, calibration, and win-rate dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/agent-performance")

    async def get_system_intelligence_institutional_memory(self) -> dict[str, Any]:
        """Get institutional-memory dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/institutional-memory")

    async def get_system_intelligence_improvement_queue(self) -> dict[str, Any]:
        """Get improvement-queue dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/improvement-queue")

    async def get_system_intelligence_anomalies(self) -> dict[str, Any]:
        """Get recent anomaly alerts for the system-intelligence dashboard."""
        return await self._client.request("GET", "/api/v1/system-intelligence/anomalies")

    async def get_system_intelligence_events(self, limit: int | None = None) -> dict[str, Any]:
        """Get recent system events for the system-intelligence dashboard."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        return await self._client.request(
            "GET", "/api/v1/system-intelligence/events", params=params
        )

    async def get_system_intelligence_km_sync(self) -> dict[str, Any]:
        """Get Knowledge Mound sync dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/km-sync")

    async def get_system_intelligence_nomic_status(self) -> dict[str, Any]:
        """Get nomic loop status dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/nomic-status")

    async def get_system_intelligence_debate_queue(self) -> dict[str, Any]:
        """Get debate queue activity dashboard data."""
        return await self._client.request("GET", "/api/v1/system-intelligence/debate-queue")

File: sdk/typescript/src/namespaces/system.ts
/**
 * System Namespace API
 *
 * Provides a namespaced interface for system administration and monitoring.
 * Includes database maintenance, history tracking, and circuit breaker management.
 */

/**
 * Maintenance task types.
 */
export type MaintenanceTask = 'status' | 'vacuum' | 'analyze' | 'checkpoint' | 'full';

/**
 * Circuit breaker status.
 */
export type CircuitBreakerStatus = 'closed' | 'open' | 'half_open';

/**
 * Cycle history entry.
 */
export interface CycleEntry {
  id: string;
  loop_id: string;
  cycle_number: number;
  started_at: string;
  completed_at?: string;
  status: string;
  debates_count: number;
  result?: unknown;
}

/**
 * Event history entry.
 */
export interface EventEntry {
  id: string;
  loop_id: string;
  event_type: string;
  timestamp: string;
  details?: unknown;
}

/**
 * Debate history entry.
 */
export interface DebateHistoryEntry {
  id: string;
  loop_id: string;
  task: string;
  status: string;
  created_at: string;
  completed_at?: string;
  consensus?: string;
  agents: string[];
}

/**
 * History summary statistics.
 */
export interface HistorySummary {
  total_debates: number;
  total_agents: number;
  total_matches: number;
}

/**
 * Circuit breaker metrics.
 */
export interface CircuitBreakerMetrics {
  status: CircuitBreakerStatus;
  failures: number;
  success_rate: number;
  last_failure?: string;
  last_success?: string;
  open_since?: string;
}

/**
 * Authentication statistics.
 */
export interface AuthStats {
  total_users: number;
  active_sessions: number;
  failed_attempts: number;
  tokens_issued: number;
  tokens_revoked: number;
}

/**
 * Maintenance result.
 */
export interface MaintenanceResult {
  task: MaintenanceTask;
  success: boolean;
  message: string;
  duration_ms?: number;
  details?: Record<string, unknown>;
}

/**
 * Debug test response.
 */
export interface DebugTestResponse {
  status: string;
  method: string;
  message: string;
}

/**
 * Response for listing cycles.
 */
export interface CyclesResponse {
  cycles: CycleEntry[];
}

/**
 * Response for listing events.
 */
export interface EventsResponse {
  events: EventEntry[];
}

/**
 * Response for listing debate history.
 */
export interface DebateHistoryResponse {
  debates: DebateHistoryEntry[];
}

/**
 * Response for circuit breakers.
 */
export type CircuitBreakersResponse = Record<string, CircuitBreakerMetrics>;

/**
 * Response for token revocation.
 */
export interface RevokeTokenResponse {
  success: boolean;
  message: string;
}

/**
 * History query options.
 */
export interface HistoryOptions {
  loop_id?: string;
  limit?: number;
}

/**
 * Options for system-intelligence event queries.
 */
export interface SystemIntelligenceEventsOptions extends Record<string, unknown> {
  limit?: number;
}

/**
 * Interface for the internal client methods used by SystemAPI.
 */
interface SystemClientInterface {
  request<T>(method: string, path: string, options?: { params?: Record<string, unknown>; body?: unknown }): Promise<T>;
}

/**
 * System API namespace.
 *
 * Provides methods for system administration:
 * - Database maintenance tasks
 * - History tracking (cycles, events, debates)
 * - Circuit breaker monitoring
 * - Authentication statistics
 * - Debug endpoints
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Run database maintenance
 * const result = await client.system.runMaintenance('vacuum');
 *
 * // Get circuit breaker status
 * const breakers = await client.system.getCircuitBreakers();
 *
 * // Get debate history
 * const { debates } = await client.system.getDebateHistory({ limit: 50 });
 *
 * // Get auth statistics
 * const stats = await client.system.getAuthStats();
 * ```
 */
export class SystemAPI {
  constructor(private client: SystemClientInterface) {}

  /**
   * Run a debug test endpoint.
   * Useful for verifying API connectivity.
   */
  async debugTest(): Promise<DebugTestResponse> {
    return this.client.request<DebugTestResponse>('POST', '/api/debug/test');
  }

  /**
   * Get cycle history.
   * @param options - Filter by loop_id and limit results
   */
  async getCycles(options?: HistoryOptions): Promise<CyclesResponse> {
    return this.client.request<CyclesResponse>('GET', '/api/history/cycles', {
      params: {
        loop_id: options?.loop_id,
        limit: options?.limit ?? 50,
      },
    });
  }

  /**
   * Get event history.
   * @param options - Filter by loop_id and limit results
   */
  async getEvents(options?: HistoryOptions): Promise<EventsResponse> {
    return this.client.request<EventsResponse>('GET', '/api/history/events', {
      params: {
        loop_id: options?.loop_id,
        limit: options?.limit ?? 100,
      },
    });
  }

  /**
   * Get debate history.
   * @param options - Filter by loop_id and limit results
   */
  async getDebateHistory(options?: HistoryOptions): Promise<DebateHistoryResponse> {
    return this.client.request<DebateHistoryResponse>('GET', '/api/history/debates', {
      params: {
        loop_id: options?.loop_id,
        limit: options?.limit ?? 50,
      },
    });
  }

  /**
   * Get history summary statistics.
   * @param loop_id - Optional filter by loop ID
   */
  async getHistorySummary(loop_id?: string): Promise<HistorySummary> {
    return this.client.request<HistorySummary>('GET', '/api/history/summary', {
      params: {
        loop_id,
      },
    });
  }

  /**
   * Run database maintenance task.
   * @param task - Type of maintenance to run
   */
  async runMaintenance(task: MaintenanceTask = 'status'): Promise<MaintenanceResult> {
    return this.client.request<MaintenanceResult>('GET', '/api/system/maintenance', {
      params: {
        task,
      },
    });
  }

  /**
   * Get authentication statistics.
   */
  async getAuthStats(): Promise<AuthStats> {
    return this.client.request<AuthStats>('POST', '/api/auth/stats');
  }

  /**
   * Revoke a token.
   * @param tokenData - Token identification data
   */
  async revokeToken(tokenData: { token_id?: string; user_id?: string }): Promise<RevokeTokenResponse> {
    return this.client.request<RevokeTokenResponse>('POST', '/api/auth/revoke', {
      body: tokenData,
    });
  }

  /**
   * Get circuit breaker metrics.
   * Returns status and metrics for all circuit breakers.
   */
  async getCircuitBreakers(): Promise<CircuitBreakersResponse> {
    return this.client.request<CircuitBreakersResponse>('POST', '/api/circuit-breakers');
  }

  /**
   * Get Prometheus metrics.
   * Returns raw Prometheus format metrics.
   */
  async getPrometheusMetrics(): Promise<string> {
    return this.client.request<string>('GET', '/metrics');
  }

  /**
   * Get high-level system intelligence dashboard stats.
   */
  async getSystemIntelligenceOverview(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/overview');
  }

  /**
   * Get agent ELO, calibration, and win-rate dashboard data.
   */
  async getSystemIntelligenceAgentPerformance(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/agent-performance');
  }

  /**
   * Get institutional-memory dashboard data.
   */
  async getSystemIntelligenceInstitutionalMemory(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/institutional-memory');
  }

  /**
   * Get improvement-queue dashboard data.
   */
  async getSystemIntelligenceImprovementQueue(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/improvement-queue');
  }

  /**
   * Get recent anomaly alerts for the system-intelligence dashboard.
   */
  async getSystemIntelligenceAnomalies(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/anomalies');
  }

  /**
   * Get recent system events for the system-intelligence dashboard.
   */
  async getSystemIntelligenceEvents(options?: SystemIntelligenceEventsOptions): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/events', {
      params: options,
    });
  }

  /**
   * Get Knowledge Mound sync dashboard data.
   */
  async getSystemIntelligenceKmSync(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/km-sync');
  }

  /**
   * Get nomic loop status dashboard data.
   */
  async getSystemIntelligenceNomicStatus(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/nomic-status');
  }

  /**
   * Get debate queue activity dashboard data.
   */
  async getSystemIntelligenceDebateQueue(): Promise<unknown> {
    return this.client.request('GET', '/api/v1/system-intelligence/debate-queue');
  }
}
